docker/release/config/opensearch/opensearch-docker-entrypoint.sh

```diff
@@ -90,22 +67,74 @@ function terminateProcesses {
     fi
 }
 
-# Enable job control so we receive SIGCHLD when a child process terminates
-set -m
+# Start up the opensearch and performance analyzer agent processes.
+# When either of them halts, this script exits, or we receive a SIGTERM or SIGINT signal then we want to kill both these processes.
+function runOpensearch {
+    # Files created by OpenSearch should always be group writable too
+    umask 0002
+
+    if [[ "$(id -u)" == "0" ]]; then
+        echo "OpenSearch cannot run as root. Please start your container as another user."
+        exit 1
+    fi
+
+    # Parse Docker env vars to customize OpenSearch
+    #
+    # e.g. Setting the env var cluster.name=testcluster
+    # will cause OpenSearch to be invoked with -Ecluster.name=testcluster
+    local opensearch_opts=()
+    while IFS='=' read -r envvar_key envvar_value
+    do
+        # OpenSearch settings need to have at least two dot separated lowercase
+        # words, e.g. `cluster.name`, except for `processors` which we handle
+        # specially
+        if [[ "$envvar_key" =~ ^[a-z0-9_]+\.[a-z0-9_]+ || "$envvar_key" == "processors" ]]; then
+            if [[ ! -z $envvar_value ]]; then
+            opensearch_opt="-E${envvar_key}=${envvar_value}"
+            opensearch_opts+=("${opensearch_opt}")
+            fi
+        fi
+    done < <(env)
+
+    setupSecurityPlugin
+
+    # Enable job control so we receive SIGCHLD when a child process terminates
+    set -m
 
-# Make sure we terminate the child processes in the event of us received TERM (e.g. "docker container stop"), INT (e.g. ctrl-C), EXIT (this script terminates for an unexpected reason), or CHLD (one of the processes terminated unexpectedly)
-trap terminateProcesses TERM INT EXIT CHLD
+    # Make sure we terminate the child processes in the event of us received TERM (e.g. "docker container stop"), INT (e.g. ctrl-C), EXIT (this script terminates for an unexpected reason), or CHLD (one of the processes terminated unexpectedly)
+    trap terminateProcesses TERM INT EXIT CHLD
 
-# Start opensearch
-$OPENSEARCH_HOME/bin/opensearch "${opensearch_opts[@]}" &
-OPENSEARCH_PID=$!
+    # Start opensearch
+    "$@" "${opensearch_opts[@]}" &
+    OPENSEARCH_PID=$!
 
-# Start performance analyzer agent
-$OPENSEARCH_HOME/bin/performance-analyzer-agent-cli > $OPENSEARCH_HOME/logs/performance-analyzer.log 2>&1 &
-PA_PID=$!
+    # Start performance analyzer agent
+    performance-analyzer-agent-cli > $OPENSEARCH_HOME/logs/performance-analyzer.log 2>&1 &
+    PA_PID=$!
 
-# Wait for the child processes to terminate
-wait $OPENSEARCH_PID
-echo "OpenSearch exited with code $?"
-wait $PA_PID
-echo "Performance analyzer exited with code $?"
+    # Wait for the child processes to terminate
+    wait $OPENSEARCH_PID
+    local opensearch_exit_code=$?
+    echo "OpenSearch exited with code ${opensearch_exit_code}"
+
+    wait $PA_PID
+    echo "Performance analyzer exited with code $?"
```

Comment:
Wonder if we want to return PA return code as well?

Reply:
Easiest way to return multiple values in bash is to rely on globals isn't it? I can do that if we have a compelling reason to, but in the meantime I don't think we absolutely need to return anything at all. To keep things consistent, I removed the return value and if it turns out useful, we can consider adding some globals to track that. Does that sound ok?

Reply:
Sounds good to me, thank you.